### PR TITLE
(API Breaking) Finish removal of block.r

### PR DESCRIPTION
**Pull request by albeanth**
## What is the change?
This follows up on #909 and fully removes the `block.r` property. Leaving it causes some (imo) unnecessary tangling. So instead of the current
```text
Reactor <--> Core <--> Assembly <--> Block <--> Component
   ^           ^                       |
   |-----------|-----------------------|
```
the following would be the new Composite relationships via properties
```text
Reactor <--> Core <--> Assembly <--> Block <--> Component
               ^                       |
               |-----------------------|
```
 

## Why is the change being made?
I can whip up some examples, but it causes some tangles with testing. Plug in the docstring it even says we should probably get rid of it. Plus you can still access the reactor object via `b.core`.

---

## Checklist

- [x] This PR has only one purpose or idea.
- [x] Tests have been added/updated to verify that the new/changed code works.

- [x] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [x] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [x] The [release notes](https://terrapower.github.io/armi/release/index.html) (location `doc/release/0.X.rst`) are up-to-date with any important changes.
- [x] The documentation is still up-to-date in the `doc` folder.
- [x] The dependencies are still up-to-date in `pyproject.toml`.

### CODE REVIEW

**Comment by albeanth** [comment]
Since this is an API change, it'll need to be run through the gamut of downstream tests.

**Comment by albeanth** [comment]
Converting to draft. This is looking like a fairly intrusive PR.

**Comment by albeanth** [comment]
The coverage report doesn't seem right again. I'm seeing the same coverage report results over several PRs. Something is fishy with coveralls. Any ideas?

**Comment by john-science** [comment]
> The coverage report doesn't seem right again. I'm seeing the same coverage report results over several PRs. Something is fishy with coveralls. Any ideas?

Fixed.

**Comment by albeanth** [comment]
See the commit message for https://github.com/terrapower/armi/pull/1425/commits/ebbcdcb6610ae5a25c7fdf3075789ca7ae105a76

TL/DR: Leaving them on the composite left the complication of `self.r.core`. Which didn't work for blocks and blocked this from going through. Since no other `Composite` other than `Block` actually ever use those methods, they can safely get moved over  the `Block` class and `self.r.core` replaced with `self.core`.

File: armi/reactor/composites.py [hunk 1]
```diff
@@ -1277,124 +1276,6 @@ def getNumberDensities(self, expandFissionProducts=False):
             return self._expandLFPs(numberDensities)
         return numberDensities
 
-    def getNeutronEnergyDepositionConstants(self):
```
[the remark this answers is not in the document]
Comment (albeanth) [reply armi/reactor/composites.py:1280]: See the commit message for https://github.com/terrapower/armi/pull/1425#issuecomment-1758274425.

TL/DR: Those methods called `self.r.core` which blocked the PR from going through. After grepping ARMI and one major downstream project, those methods never get called on any `Composite` other than `Block`. So those methods can safely be moved over to the `Block` class and `self.r.core` can get replaced with `self.core`.

**Comment by john-science** [comment]
To be clear, the reason I haven't tried to merge this is that we are currently in a code freeze. I am not ignoring you.

**Comment by john-science** [comment]
@albeanth  This PR is getting a bit old; there are a bunch of conflicts.

Also, we should move the release note from `0.2.9` to `0.3.1`.

**Comment by albeanth** [comment]
> This PR is getting a bit old; there are a bunch of conflicts.
> 
> Also, we should move the release note from `0.2.9` to `0.3.1`.

@john-science ok, the merge conflicts have been resolved and this feature branch is now up to date with `main`

File: armi/reactor/blocks.py [hunk 2]
```diff
@@ -176,42 +177,6 @@ def core(self):
         c = self.getAncestor(lambda c: isinstance(c, Core))
         return c
```
Comment (john-science) [inline armi/reactor/blocks.py:178]: It's still weird that we have `Block.core`, but fine. That's not for this PR.

Also... there is no error handling on `Block.core`? Mark that as sus.

File: armi/reactor/composites.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (john-science) [reply armi/reactor/composites.py:1354]: Okay!  As long as we call this out in the "API changes" section of the release notes!

File: doc/release/0.2.rst [hunk 3]
```diff
@@ -1,6 +1,37 @@
 ***********************
-ARMI v0.2 Release Notes
+ARMI v0.3 Release Notes
+=======================
+
+ARMI v0.3.1
+============
+Release Date: TBD
```
Comment (john-science) [inline doc/release/0.2.rst:7]: Whoops!

It looks like you accidentally over-wrote our `0.2` release notes file.  We have a totally different file for the `0.3` release notes.

I don't want to LOSE the old release notes, can you move this to the existing `doc/0.3.rst` files?

Thanks!

File: armi/reactor/blocks.py (hunk 2 above)
Comment (albeanth) [reply armi/reactor/blocks.py:177]: What sort of error handling do you have in mind?

Comment (john-science) [reply armi/reactor/blocks.py:177]: What if the `Block` is not part of a `Core`?

You know, when you're building a `Reactor` from `Blueprints`, you have to create a `Block` and add it to the `Core`.  Or you might be doing some fuel shuffling, and not all `Assemblies`/`Blocks` are in the `Core` at the same time.  Or your `Block` might be in the `SpentFuelPool`.  Lots of situations.

But the question is... should that return `None`, or `raise Error()` or what?  Hmm... What works, but isn't heavy handed?  hmmm

Comment (albeanth) [reply armi/reactor/blocks.py:177]: Hmmmm. Well `getAncestor` currently returns `None`. I suppose for the case where it's a SFP or something else, I could see a runLog or something being useful. 

Though resolving this feels like worthy of a separate PR. What do you think?

File: doc/release/0.2.rst (hunk 3 above)
Comment (albeanth) [reply doc/release/0.2.rst]: Whoops, fixed! https://github.com/terrapower/armi/pull/1425/commits/c2cd59d54051d93b744c0c49e7912eab6b5458c1

File: armi/reactor/composites.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (albeanth) [reply armi/reactor/composites.py:1354]: https://github.com/terrapower/armi/pull/1425/commits/c2cd59d54051d93b744c0c49e7912eab6b5458c1

**Review by john-science — APPROVED** [review]

**Pull request merged**